### DISCUSSION TITLE
[feature] Batch-collect

### DISCUSSION THREAD

**Discussion by unindented** [umami-software/umami#1473] [General]
When an app needs to send many events, it'd be desirable to have some way of batching them, instead of having to make a `POST` request for each.

I'm thinking of either modifying `POST /api/collect` to also accept an array of `pageview` and `event` payloads, or creating a new `POST /api/collect/batch` endpoint.

It could impose limits, like Google Analytics does: https://developers.google.com/analytics/devguides/collection/protocol/v1/devguide#batch

What do you think?

**Comment by mikecao** [comment]
The problem is we don't know if a user will return or will trigger any events. So if we wait for a batch to send, we could lose that data.

**Comment by unindented** [comment]
I'm using the API directly, not via `umami.js`. I persist events in localStorage/IndexedDB, and send them whenever the user goes online. Making one request per event in that scenario is very inefficient.

Google Analytics supports batching for that reason, among others: https://developers.google.com/analytics/devguides/reporting/core/v3/batching

**Comment by mikecao** [comment]
I see what you mean. Yes we should support that use case for the API.

**Comment by astappiev** [comment]
Any ETA on this? I would like to have such an endpoint.

**Comment by zeroCoder1** [comment]
I’m using this API for native mobile apps and would love to use the batching feature. Could you please let me know when it’s expected to be available? Thanks a bunch!

**Comment by zeroCoder1** [comment]
https://github.com/umami-software/umami/pull/3275

WDYT?
@unindented @mikecao @astappiev

**Comment by astappiev** [comment]
Thank you @zeroCoder1, that's exactly what I need
